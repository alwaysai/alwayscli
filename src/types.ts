import { BRANCH, LEAF } from './constants';

export type Argv<R extends boolean> = R extends true ? string[] : (string[] | undefined);

type GetValueSync<V, R extends boolean> = (argv: Argv<R>) => V;
type GetValueAsync<V, R extends boolean> = (argv: Argv<R>) => Promise<V>;
type GetValue<V, R extends boolean> = GetValueSync<V, R> | GetValueAsync<V, R>;

type AnyGetValue = GetValue<any, any>;

export type Input<V, R extends boolean = boolean> = {
  required?: R;
  getValue: GetValue<V, R>;
  getDescription?: () => string | undefined;
  placeholder: string;
};

export type AnyInput = Input<any>;

export type InputValue<T extends AnyGetValue> = T extends GetValueAsync<infer U, any>
  ? U
  : T extends GetValueSync<infer U, any>
  ? U
  : never;

export type AnyOptions = {
  [argName: string]: AnyInput;
};

export type NamedValues<T extends AnyOptions> = {
  [K in keyof T]: InputValue<T[K]['getValue']>
};

export type Command = Branch | Leaf<AnyInput, AnyOptions>;

export type Branch = {
  _type: typeof BRANCH;
  name: string;
  description?: string;
  subcommands: (Branch | Leaf<any, any>)[];
};

export type Leaf<T extends AnyInput, U extends AnyOptions> = {
  _type: typeof LEAF;
  name: string;
  description?: string;
  args?: T;
  options?: U;
  action: (args: InputValue<T['getValue']>, options: NamedValues<U>) => any;
};

// The "commandType" field is assigned internally by the framework.
// This helper function is used to remove that field for user use.
export type ExcludeUnderscoreType<T extends { _type: any }> = Pick<
  T,
  Exclude<keyof T, '_type'>
>;

export type CommandStack = {
  branches: Branch[];
  leaf?: Leaf<AnyInput, any>;
};
