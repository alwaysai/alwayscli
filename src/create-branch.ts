import { Branch, ExcludeUnderscoreType } from './types';
import { BRANCH } from './constants';

type Config = ExcludeUnderscoreType<Branch>;

export function createBranch(config: Config) {
  const branch: Branch = {
    ...config,
    _type: BRANCH,
  };
  return branch;
}
