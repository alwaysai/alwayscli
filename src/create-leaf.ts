import { AnyOptions, Leaf, ExcludeUnderscoreType, AnyInput } from './types';
import { LEAF } from './constants';

export function createLeaf<T extends AnyInput = AnyInput, U extends AnyOptions = {}>(
  config: ExcludeUnderscoreType<Leaf<T, U>>,
) {
  const leaf: Leaf<T, U> = {
    ...config,
    _type: LEAF,
  };
  return leaf;
}
