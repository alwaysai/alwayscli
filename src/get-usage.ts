import redent = require('redent');
import chalk from 'chalk';

import { Command } from './types';
import { BRANCH, LEAF } from './constants';
import { createTextList } from './create-text-list';
import { regularizeText, wrapInSquareBrackets } from './util';

const INDENT_SIZE = 3;

const RED_ERROR = chalk.red('Error:');

export function getUsage(commands: Command[], errorMessage?: string) {
  const command = commands.slice(-1)[0];
  // ^^ Last command on the stack. Could be leaf or branch.

  let firstParagraph = `Usage: ${commands.map(command => command.name).join(' ')}`;

  const otherParagraphs: string[] = [];
  switch (command._type) {
    case BRANCH:
      firstParagraph += ' <subcommand> ...';
      otherParagraphs.push('Subcommands:');
      const items: Parameters<typeof createTextList> = command.subcommands.map(
        ({ name, description }) => ({
          name,
          text: description,
        }),
      );
      const subcommandsParagraph = redent(createTextList(...items), INDENT_SIZE);
      otherParagraphs.push(subcommandsParagraph);
      break;
    case LEAF:
      const { args: positionalInput, options: namedInputs } = command;
      if (positionalInput) {
        const { placeholder, getDescription, required } = positionalInput;
        firstParagraph += ` ${
          required ? placeholder : wrapInSquareBrackets(placeholder)
        }`;
        const description = getDescription && getDescription();
        if (description) {
          otherParagraphs.push(`${placeholder}:`);
          otherParagraphs.push(redent(regularizeText(description), INDENT_SIZE));
        }
      }
      if (namedInputs) {
        const entries = Object.entries(namedInputs);
        if (entries.length > 0) {
          const optionsNotRequired = entries.every(
            ([_, namedInput]) => !namedInput.required,
          );
          firstParagraph += optionsNotRequired ? ' [<options>]' : ' <options>';
          otherParagraphs.push('Options:');
          const items: Parameters<typeof createTextList> = entries.map(
            ([inputName, input]) => {
              let name = `--${inputName}`;
              if (input.placeholder) {
                name += ` ${input.placeholder}`;
              }
              if (!input.required) {
                name = wrapInSquareBrackets(name);
              }
              const text = input.getDescription && input.getDescription();
              return { name, text };
            },
          );
          const optionsParagraph = redent(createTextList(...items), INDENT_SIZE);
          otherParagraphs.push(optionsParagraph);
        }
      }
      break;
    default:
      throw new Error('Unexpected command type');
  }

  const paragraphs = [firstParagraph];

  const regularizedDescription = regularizeText(command.description);
  if (regularizedDescription) {
    paragraphs.push(redent(regularizedDescription, INDENT_SIZE));
  }

  paragraphs.push(...otherParagraphs);

  if (errorMessage) {
    paragraphs.push(`${RED_ERROR} ${errorMessage}`);
  }

  return paragraphs.join('\n\n');
}
