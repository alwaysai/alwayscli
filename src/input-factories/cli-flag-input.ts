import { CliInput } from '../types';
import { UsageError } from '../usage-error';

type Config = { description?: string; hidden?: boolean };

export function CliFlagInput(config: Config = {}) {
  const { description, hidden } = config;
  const input: CliInput<boolean, false> = {
    placeholder: '',
    required: false,
    hidden,
    getValue(argv) {
      if (!argv) {
        return false;
      }
      if (argv.length !== 0) {
        throw new UsageError('Flag input cannot receive values');
      }
      return true;
    },
    description,
  };
  return input;
}
