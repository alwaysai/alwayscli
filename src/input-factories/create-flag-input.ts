import { Input } from '../types';
import { UsageError } from '../usage-error';

export function createFlagInput(config: { description?: string } = {}) {
  const input: Input<boolean, false> = {
    getValue(argv) {
      if (!argv) {
        return false;
      }
      if (argv.length !== 0) {
        throw new UsageError('Flag input cannot receive values');
      }
      return true;
    },
    getDescription() {
      return config.description;
    },
    placeholder: '',
    required: false,
  };
  return input;
}
