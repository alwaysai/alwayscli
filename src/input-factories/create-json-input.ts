import parseJson = require('parse-json');

import { Input } from '../types';
import { UsageError } from '../usage-error';

type Config = { description?: string; required?: boolean; placeholder?: string };

// Because a JSON input value has type `any`, we don't need to do
// anything fancy with function overloads to handle the "required"
// field like we do for some other input factories.
export function createJsonInput(config: Config = {}) {
  const { placeholder = '<json>', required, description } = config;
  const input: Input<any> = {
    required,
    placeholder,
    getValue(argv) {
      if (!argv) {
        return;
      }
      if (argv.length !== 1) {
        throw new UsageError(`Expected a single ${placeholder} string`);
      }
      try {
        const parsed = parseJson(argv[0]);
        return parsed;
      } catch (ex) {
        throw new UsageError(ex.message || 'Failed to parse JSON');
      }
    },
    getDescription() {
      return description;
    },
  };
  return input;
}
