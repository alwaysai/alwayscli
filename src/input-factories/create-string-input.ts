import { Input } from '../types';
import { UsageError } from '../usage-error';

type Config = Partial<{
  required: boolean;
  description: string;
  defaultValue: string;
  placeholder?: string;
}>;

const PLACEHOLDER = '<str>';

export { createStringInput };
function createStringInput(
  config: Config & { defaultValue: string },
): Input<string, false>;
function createStringInput(config: Config & { required: true }): Input<string, true>;
function createStringInput(config?: Config): Input<string | undefined, false>;
function createStringInput(config: Config = {}) {
  const { defaultValue, required, description, placeholder = PLACEHOLDER } = config;
  const input: Input<string | undefined> = {
    getValue(argv) {
      if (!argv) {
        return typeof defaultValue === 'string' ? defaultValue : undefined;
      }

      if (argv.length > 1) {
        throw new UsageError(`Expected just one ${PLACEHOLDER}`);
      }

      if (argv.length === 0) {
        throw new UsageError(`Expected a ${PLACEHOLDER}`);
      }

      return argv[0];
    },
    getDescription() {
      return description;
    },
    placeholder,
    required,
  };
  return input;
}
