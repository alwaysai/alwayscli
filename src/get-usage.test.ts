import { getUsage } from './get-usage';
import { createBranch } from './create-branch';
import { createStringInput } from './input-factories/create-string-input';
import { Leaf } from './types';
import { createLeaf } from './create-leaf';

const leaf = createLeaf({
  name: 'echo',
  options: {
    message: createStringInput({ description: 'A message' }),
  },
  action(foo) {
    return foo;
  },
});

const root = createBranch({
  name: 'cli',
  subcommands: [leaf],
});

describe(getUsage.name, () => {
  it('Creates a usage string for a branch', () => {
    const usageString = getUsage([root]);
    expect(usageString).toMatchSnapshot();
  });

  it('Creates a usage string for a leaf', () => {
    const usageString = getUsage([root, leaf as Leaf<any, any, any>]);
    expect(usageString).toMatchSnapshot();
  });
});
