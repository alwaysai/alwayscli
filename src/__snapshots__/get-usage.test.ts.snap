// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUsage Creates a usage string for a branch 1`] = `
"Usage: cli <subcommand> ...

Subcommands:

   echo"
`;

exports[`getUsage Creates a usage string for a leaf 1`] = `
"Usage: cli echo [<options>]

Options:

   [--message <str>] : A message"
`;
